### PR TITLE
Bug 1122423 [Stingray] UI-Transition - Deck launcher waving effect

### PR DESCRIPTION
**Pull request by sean2449**

### CODE REVIEW

**Comment by try-server-hook** [comment]
[sean2449 Sung Chiu (sean2449)](https://github.com/sean2449) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=e9280ecbdb52)

File: tv_apps/smart-home/js/home.js [hunk 1]
```diff
@@ -165,6 +165,46 @@
...
 
+    _fillDeckIcon: function(cardButton, card) {
+
+      // deck's icon using gaia font
+      var deckIcon = document.createElement('span');
+      deckIcon.className = 'icon';
+      deckIcon.dataset.icon = card.fontIcon;
+
+      // front wave of a deck
+      var waveFront = document.createElement('div');
```
Comment (huchengtw-moz) [inline tv_apps/smart-home/js/home.js]: How about create a method call "createWave"?
The lines creating wave are totally the same but different in class names.

File: tv_apps/tv_shared/components/smart-button/script.js [hunk 2]
```diff
@@ -42,6 +42,17 @@ window.SmartButton = (function(win) {
         }
         break;
       case 'transitionend':
+        // fire the event to trigger wave effect
+        // there are many transition properties in focus
+        // we only want to fire one event
+        if (this.classList.length === 2 &&
```
Comment (huchengtw-moz) [inline tv_apps/tv_shared/components/smart-button/script.js]: This condition is not accurate. Users may add more than 2 class to class list. In this case, the event is broken.

File: tv_apps/tv_shared/js/cards/deck.js [hunk 3]
```diff
@@ -6,6 +6,10 @@
   var Deck = function Deck(options) {
     this.nativeApp = options.nativeApp;
     this.name = options.name;
+    this.fontIcon = options.fontIcon;
```
Comment (huchengtw-moz) [inline tv_apps/tv_shared/js/cards/deck.js]: Since cards are shared by app-deck and smart-home, you should ask dwi2 to sync about this.

**Comment by try-server-hook** [comment]
[sean2449 Sung Chiu (sean2449)](https://github.com/sean2449) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=3e319b41f514)

File: tv_apps/smart-home/style/deck.css [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+  display: inline-block;
+  line-height: 28rem;
+}
+
+smart-button .icon[data-icon]:before {
+  font-size: 30rem;
+}
+
+smart-button[type="app-button"].deck-tv {
+  background-color: #D70270;
```
Comment (huchengtw-moz) [inline tv_apps/smart-home/style/deck.css]: there is a whitespace at the end of this line.

File: tv_apps/smart-home/style/deck.css [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+smart-button .icon[data-icon]:before {
+  font-size: 30rem;
+}
+
+smart-button[type="app-button"].deck-tv {
+  background-color: #D70270; 
+}
+
+smart-button[type="app-button"].deck-application {
+  background-color: #07CAB0;
```
Comment (huchengtw-moz) [inline tv_apps/smart-home/style/deck.css]: the same here.

File: tv_apps/smart-home/style/deck.css [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+smart-button[type="app-button"].deck-tv {
+  background-color: #D70270; 
+}
+
+smart-button[type="app-button"].deck-application {
+  background-color: #07CAB0; 
+}
+
+smart-button[type="app-button"].deck-device {
+  background-color: #0074D4;
```
Comment (huchengtw-moz) [inline tv_apps/smart-home/style/deck.css]: the same here.

File: tv_apps/smart-home/style/deck.css [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+smart-button[type="app-button"].deck-application {
+  background-color: #07CAB0; 
+}
+
+smart-button[type="app-button"].deck-device {
+  background-color: #0074D4; 
+}
+
+smart-button[type="app-button"].deck-dashboard {
+  background-color: #FF7E00;
```
Comment (huchengtw-moz) [inline tv_apps/smart-home/style/deck.css]: the same here.

**Comment by try-server-hook** [comment]
[sean2449 Sung Chiu (sean2449)](https://github.com/sean2449) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=343d16fbd92b)

**Comment by try-server-hook** [comment]
[sean2449 Sung Chiu (sean2449)](https://github.com/sean2449) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=82fc99ed5aa6)

File: tv_apps/smart-home/style/deck.css [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+
+smart-button .icon[data-icon] {
+  position: relative;
+  color: white;
+  display: inline-block;
+  line-height: 28rem;
+}
+
+smart-button .icon[data-icon]:before {
+  font-size: 30rem;
```
Comment (rexboy7) [inline tv_apps/smart-home/style/deck.css]: Why do we need to set such big font-size?

Comment (rexboy7) [inline tv_apps/smart-home/style/deck.css]: oh I got it, so it's icon for deck.

File: tv_apps/smart-home/style/deck.css [hunk 9]
```diff
@@ -0,0 +1,155 @@
...
+/**
+ * Deck wave effect
+ */
+
+.deck-wave {
+  position: absolute;
+  width: 134.4rem;
+  height: 67.2rem;
+  transform: scale(1.2, 1.2);
+  animation: defaultAnimation 5.0s linear 0s infinite normal;
```
Comment (rexboy7) [inline tv_apps/smart-home/style/deck.css]: Please change to a more specific name than `defaultAnimation`.

File: tv_apps/smart-home/js/home.js [hunk 10]
```diff
@@ -165,6 +165,44 @@
...
+      cardButton.addEventListener('blur', function(evt) {
+          waveBack.classList.add('wave-paused');
+          waveFront.classList.add('wave-paused');
+      });
+
+      cardButton.appendChild(waveBack);
+      cardButton.appendChild(deckIcon);
+      cardButton.appendChild(waveFront);
+      cardButton.classList.add('deck-' + card.deckClass);
+    },
```
Comment (rexboy7) [inline tv_apps/smart-home/js/home.js]: I think it's about time to split out logics of drawing animation and icons for cards, but it's ok for this patch. We can file another bug for refactoring.

File: tv_apps/smart-home/style/deck.css [hunk 11]
```diff
@@ -0,0 +1,155 @@
...
+
+@keyframes dashboardWaveFrontAnimation {
+  from {left: -35rem;}
+  to {left: -115.64rem;}
+}
+
+@keyframes dashboardWaveBackAnimation {
+  from {left: -32rem;}
+  to {left: -112.64rem;}
+}
```
Comment (rexboy7) [inline tv_apps/smart-home/style/deck.css]: add a newline here

File: tv_apps/smart-home/style/deck.css [hunk 12]
```diff
@@ -0,0 +1,155 @@
...
+
+/**
+ * Deck wave effect
+ */
+
+.deck-wave {
+  position: absolute;
+  width: 134.4rem;
+  height: 67.2rem;
+  transform: scale(1.2, 1.2);
```
Comment (rexboy7) [inline tv_apps/smart-home/style/deck.css]: `scale(1.2)` would do the work.

File: tv_apps/smart-home/style/deck.css [hunk 13]
```diff
@@ -0,0 +1,155 @@
...
+
+.wave-front {
+  animation-duration: 3.05s
+}
+
+.wave-back {
+  animation-duration: 3.0s;
+}
+
+div.wave-paused{
```
Comment (rexboy7) [inline tv_apps/smart-home/style/deck.css]: add a whitespace here

**Comment by try-server-hook** [comment]
[sean2449 Sung Chiu (sean2449)](https://github.com/sean2449) started tests. [Results](https://treeherder.mozilla.org/ui/#/jobs?repo=gaia-try&revision=daaf59f48c52)

**Pull request merged**